Latency-friendly LV modal implementation and Goals form refactor

zoldar opened the pull request:
### Changes

This PR:

- implements universal LV modal component ([documentation](https://github.com/plausible/analytics/pull/3649/files#diff-2da12bc470c3b45efacd72241d0c957fb5bad2c49aa0d87d3822bf19487070a1R2-R84))
- refactor ComboBox compoenent to provide better feedback on high latency connections
- refactors Goal Settings form to use the new modal and makes the form contents provide better feedback on high latency connections

### Tests
- [x] Automated tests have been added
- [ ] This PR does not require tests

### Changelog
- [ ] Entry has been added to changelog
- [x] This PR does not make a user-facing change

### Documentation
- [ ] [Docs](https://github.com/plausible/docs) have been updated
- [x] This change does not need a documentation update

### Dark mode
- [x] The UI has been tested both in dark and light mode
- [ ] This PR does not change the UI

zoldar commented on assets/js/app.js:
```diff
@@ -10,6 +10,16 @@ Alpine.data('dropdown', dropdown)
 Alpine.start()
 
+if (document.querySelectorAll('[data-modal]').length > 0) {
+  window.addEventListener(`phx:close-modal`, (e) => {
```
This event listener allows hiding modal from the backend using `Modal.close/2`

zoldar commented on extra/lib/plausible_web/live/funnel_settings/form.ex:
```diff
@@ -84,6 +84,9 @@ defmodule PlausibleWeb.Live.FunnelSettings.Form do
                     submit_name="funnel[steps][][goal_id]"
                     module={PlausibleWeb.Live.Components.ComboBox}
                     suggest_fun={&PlausibleWeb.Live.Components.ComboBox.StaticSearch.suggest/2}
+                    on_selection_made={fn value, by_id -> 
+                      send(self(), {:selection_made, %{submit_value: value, by: by_id}}) 
+                    end}
```
A problem with implicit send surfaced while working on the modal. As this message is used only in some uses of ComboBox, it made sense to make it optional, so that caller is not forced to implement noop `handle_info` clause just for that. Additionally, callback approach allows for greater flexibility.

zoldar commented on lib/plausible_web/live/components/modal.ex:
```diff
@@ -0,0 +1,80 @@
+    end
+  end
+
+  def close(socket, id) do
+    Phoenix.LiveView.push_event(socket, "close-modal", %{id: id})
+  end
```
Closing the modal outside normal input events is largely something initiated from the backend, so `close/2` is exposed as the only way to do that programmatically for now.

zoldar commented on lib/plausible_web/live/components/modal.ex:
```diff
@@ -0,0 +1,80 @@
+    ~H"""
+    <div
+      id={@id}
+      class="[&[data-phx-ref]_div.modal-dialog]:hidden [&[data-phx-ref]_div.modal-loading]:block"
```
CSS assures immediate loading state application the moment state change is dispatched to the backend. It also lets us avoid "flicker" issues when modal is rapidly closed and opened again.

zoldar commented on lib/plausible_web/live/components/modal.ex:
```diff
@@ -0,0 +1,80 @@
+      x-cloak
+      x-data="{ 
+        modalOpen: false, 
+        openModal() {
+          liveSocket.execJS($el, $el.dataset.onopen);
+          this.modalOpen = true;
+        },
+        closeModal() { 
+          this.modalOpen = false; 
+          liveSocket.execJS($el, $el.dataset.onclose);
+        }
```
This way of "duplicated" management of modal state allows for responsiveness while assuring full reload of contents when needed.

zoldar commented on lib/plausible_web/live/components/modal.ex:
```diff
@@ -0,0 +1,80 @@
+        <div :if={@load_content?} class="modal-dialog w-1/2 h-full" x-on:click.outside="closeModal()">
+          <%= render_slot(@inner_block) %>
+        </div>
+        <div class="modal-loading hidden w-1/2 h-full">
+          <div class="text-center max-w-md w-full mx-auto bg-white dark:bg-gray-800 shadow-md rounded px-8 pt-6 pb-8 mb-4 mt-8">
```
Styling will most likely have to be made customizable here, so that loading status at least preserves the width of the loaded modal.

zoldar commented on lib/plausible_web/live/goal_settings/form.ex:
```diff
@@ -2,85 +2,71 @@ defmodule PlausibleWeb.Live.GoalSettings.Form do
   Live view for the goal creation form
   """
-  use Phoenix.LiveView
+  use Phoenix.LiveComponent
```
Turned it into a LiveComponent in the process while iterating on previous approaches to responsive modal. It does not have to be a full-fledged liveview anyway and it also slightly simplifies logic IMHO.

zoldar commented on lib/plausible_web/live/goal_settings/form.ex:
```diff
@@ -2,85 +2,71 @@ defmodule PlausibleWeb.Live.GoalSettings.Form do
+        <h2 class="text-xl font-black dark:text-gray-100">Add Goal for <%= @domain %></h2>
 
-          <.tabs tabs={@tabs} />
+        <.tabs tabs={@tabs} myself={@myself} />
```
Tabs will need changing with either an explicit loading state when switching or managing them on frontent entirely. Both ways are viable I think. Anyways, that's a bit beside the scope of that experiment but should be addressed eventually.

zoldar commented on lib/plausible_web/live/goal_settings/form.ex:
```diff
@@ -282,19 +270,18 @@ defmodule PlausibleWeb.Live.GoalSettings.Form do
+        socket =
+          goal
+          |> Map.put(:funnels, [])
+          |> socket.assigns.on_save_goal.(socket)
```
Implementing callback here for better flexibility.

zoldar commented on lib/plausible_web/live/goal_settings.ex:
```diff
@@ -39,19 +41,21 @@ defmodule PlausibleWeb.Live.GoalSettings do
+          site={@site}
+          current_user={@current_user}
+          on_save_goal={
+            fn goal, socket ->
+              send(self(), {:goal_added, goal})
+              Modal.close(socket, "goals-form-modal")
+            end
```
Callback enables greater flexibility. Also, the form component is now completely agnostic to the context it is put in and matters of handling added goal and managing modal state are kept separate.

zoldar commented on lib/plausible_web/live/components/modal.ex:
```diff
@@ -0,0 +1,89 @@
+      x-data="{
+        modalOpen: false,
+        openModal() {
+          liveSocket.execJS($el, $el.dataset.onclose);
+          liveSocket.execJS($el, $el.dataset.onopen);
```
Sending those events consecutively one after another allows to consistently reset modal state while avoiding race conditions.

aerosol commented on lib/plausible_web/live/components/modal.ex:
```diff
@@ -0,0 +1,177 @@
+      to ensure that.
+    * `Modal.close/2` - to close the modal from the backend; usually
+      done inside wrapped component's `handle_event/2`. The example
+      qouted above shows one way to implement this, under that assumption
```
```suggestion
      quoted above shows one way to implement this, under that assumption
```

aerosol reviewed (approved):
Looks good, two minor issues:

- clicking outside of the modal in the highlighted area feels random - sometimes does close the modal, sometimes it doesn't

![image](https://github.com/plausible/analytics/assets/173738/142e7d0d-244a-4177-9ab4-653229d90e0c)

- Since the selection messages are unnecessary, we can also revert the test change done at https://github.com/plausible/analytics/pull/3675/commits/46f65d9adb9dd220f6043e6589fc48865a905710

zoldar commented:
@aerosol I have fixed and cleaned up modal styling, especially on small/mobile screens. It also addresses problem with active areas for clicking "outside" the dialog window.

Before (what's currently in prod too 🙈 ):

<img width="393" alt="image" src="https://github.com/plausible/analytics/assets/588351/a9446e00-c76d-4a33-ad7f-ce7bb2dabc9d">

After:

<img width="399" alt="image" src="https://github.com/plausible/analytics/assets/588351/c6d4784b-a73b-44d7-a95c-82e93c8b7633">

Done in https://github.com/plausible/analytics/pull/3649/commits/3664027671546148d5c5676290b6e5016375e0c2

zoldar commented:
Following changes were applied:

* Show a spinner directly on backdrop instead of in a dialog
* Both spinner and dialog centered vertically in the viewport
* Scroll should be trapped within the modal (like details views on the dashboard)
* Focus should be trapped within the modal (Phoenix.Component.focus_wrap/1)
* Enter/leave transitions, shadow, border radius from https://tailwindui.com/components/application-ui/overlays/dialogs#component-a522f43bd52fbf6391e2273163eec330

aerosol reviewed (changes requested):
There's a bit inconsistent behavior now with regards to the modal placement - in goals, it shows exactly in the center now:

![image](https://github.com/plausible/analytics/assets/173738/0d0991d0-beee-4dd3-9253-0b040937fc6c)

vs custom props for example:

![image](https://github.com/plausible/analytics/assets/173738/1046a848-e763-4783-9ccd-e2b1b932cd81)

This dynamic (?) alignment makes it look glitchy when switching tabs:


https://github.com/plausible/analytics/assets/173738/3fdfa880-a3e8-4ff7-a670-04840208c96b

zoldar commented:
https://github.com/plausible/analytics/pull/3649/commits/653a256d14b05bdf91258c1535d31b42481f7425 brings back top-alignment.

Before:

https://github.com/plausible/analytics/assets/588351/e59ad3ab-a400-41a5-b806-88c65b8a1601

After:


https://github.com/plausible/analytics/assets/588351/39c612de-0b5c-46b7-ba8d-d06edfc63a5d
